python/tritonserver/_api/_tensor.py

```diff
@@ -0,0 +1,499 @@
+
+        tensor = Tensor.from_dlpack(torch.zeros(100, dtype=torch.float16))
+
+        """
+        dlpack_object = DLPackObject(obj)
+        data_type = dlpack_object.triton_data_type
+        shape = dlpack_object.shape
+        memory_buffer = MemoryBuffer._from_dlpack_object(
+            obj, dlpack_object=dlpack_object
+        )
+        return Tensor(data_type, shape, memory_buffer)
+
+    def to_host(self) -> Tensor:
+        """Move the tensor to CPU memory from device memory
+
+        Returns
+        -------
+        Tensor
+            The tensor moved to the CPU.
+
+        Examples
+        --------
+
+        tensor = Tensor.from_dlpack(torch.zeros(100, dtype=torch.float16).to("cuda"))
+
+        numpy_nd_array = numpy.array(tensor.to_host())
+
+        """
+        return self.to_device("cpu")
+
+    def to_device(self, device: DeviceOrMemoryType) -> Tensor:
+        """Move the tensor to the specified device.
+
+        Parameters
+        ----------
+        device : DeviceOrMemoryType
+            The target device. Device can be specified as a string,
+            MemoryType, tuple [MemoryType, memory_type__id], or
+            tuple[DLDeviceType, device_id].
+
+        Returns
+        -------
+        Tensor
+            The tensor moved to the specified device.
+        """
+        memory_type, memory_type_id = parse_device_or_memory_type(device)
+        if self.memory_type == memory_type and self.memory_type_id == memory_type_id:
+            return self
+        if self.memory_type == MemoryType.CPU_PINNED and memory_type == MemoryType.CPU:
+            return self
+        if cupy is not None:
+            if self.memory_type in (MemoryType.CPU, MemoryType.CPU_PINNED):
+                ndarray = numpy.from_dlpack(self)
+            else:
+                ndarray = cupy.from_dlpack(self)
+
+            if memory_type == MemoryType.CPU:
+                return Tensor.from_dlpack(cupy.asnumpy(ndarray))
+            if memory_type == MemoryType.GPU:
+                with cupy.cuda.Device(memory_type_id):
+                    return Tensor.from_dlpack(cupy.asarray(ndarray))
+
+        raise UnsupportedError(
+            f"Conversion from {(self.memory_type,self.memory_type_id)} to {(memory_type, memory_type_id)} not supported."
+        )
+
+    def _to_numpy_on_host(self) -> numpy.ndarray:
+        if self.memory_type in (MemoryType.CPU, MemoryType.CPU_PINNED):
+            return numpy.from_dlpack(self)
+
+        if cupy is not None:
+            return cupy.asnumpy(cupy.from_dlpack(self))
+
+        raise UnsupportedError(
+            f"Conversion from {self.memory_type} to numpy array not supported."
+        )
+
+    @staticmethod
+    def _deserialize_bytes_array(numpy_ndarray: numpy.ndarray) -> numpy.ndarray:
+        result = []
+        _buffer = memoryview(numpy_ndarray)
+        offset = 0
+        while offset < len(_buffer):
+            (item_length,) = struct.unpack_from("@I", _buffer, offset)
```

Comment:
note: endianness if this doesn't match with our decisions around client/server endian spots